app/src/pages/playground/usePlaygroundPrompt.tsx

```diff
@@ -0,0 +1,211 @@
+            }))
+          : [],
+    },
+    tools: promptVersion.tools.map((t) => ({
+      id: generateToolId(),
+      definition: t.definition,
+    })),
+  } satisfies Partial<PlaygroundInstance>;
+};
+
+export const instanceToPromptVersion = (instance: PlaygroundInstance) => {
+  if (!instance.prompt) {
+    return null;
+  }
+  if (instance.template.__type === "text_completion") {
+    // eslint-disable-next-line no-console
+    console.warn(
+      "Instance to prompt version conversion not supported for text completion"
+    );
+    return null;
+  }
+  return {
+    id: instance.prompt.id,
+    modelName: instance.model.modelName || DEFAULT_MODEL_NAME,
+    modelProvider: instance.model.provider,
+    templateType: "CHAT",
+    template: {
+      __typename: "PromptChatTemplate",
+      messages: instance.template.messages.map((m) => ({
+        content: m.content,
+        role: chatMessageRoleToPromptMessageRole(m.role),
+      })),
+    },
+    tools: instance.tools.map((t) => ({
+      __typename: "ToolDefinition",
+      definition: t.definition,
+    })),
+    // @TODO(apowell): Add description and invocationParameters
+  } satisfies Partial<PromptVersion>;
+};
+
+const getLatestPromptVersion = (
+  prompt: usePlaygroundPromptQuery$data["prompt"]
+) => {
+  if (!prompt) {
+    return null;
+  }
+  return prompt?.promptVersions?.edges?.[0]
+    ?.promptVersion as Mutable<PromptVersion> | null;
+};
+
+const query = graphql`
+  query usePlaygroundPromptQuery($promptId: GlobalID!) {
+    prompt: node(id: $promptId) {
+      ... on Prompt {
+        id
+        name
+        createdAt
+        description
+        promptVersions(first: 1) {
+          edges {
+            promptVersion: node {
+              id
+              description
+              modelName
+              modelProvider
+              invocationParameters
+              templateType
+              template {
+                __typename
+                ... on PromptChatTemplate {
+                  messages {
+                    ... on TextPromptMessage {
+                      content
+                      role
+                    }
+                  }
+                }
+              }
+              tools {
+                __typename
+                definition
+              }
+            }
+          }
+        }
+      }
+    }
+  }
+`;
+
+export const usePlaygroundPromptReference = (instanceId: number) => {
+  const instances = usePlaygroundContext((state) => state.instances);
+  const instance = instances.find((instance) => instance.id === instanceId);
+  if (!instance) {
+    throw new Error(`Instance ${instanceId} not found`);
+  }
+  const promptId = instance.prompt?.id;
+  const [queryReference, loadQuery, disposeQuery] =
+    useQueryLoader<usePlaygroundPromptQuery>(query);
+
+  useEffect(() => {
+    if (promptId) {
+      loadQuery({ promptId: promptId });
+      return () => disposeQuery();
+    }
+  }, [disposeQuery, loadQuery, promptId]);
```

Comment:
I personally am not really a fan of this pattern - the preloaded queries are really designed to preload queries that can be loaded / disposed on UI interaction (like switching tabs or clicking a button).

Can't this be largely simplified if fetch query is just used? e.x. https://github.com/Arize-ai/phoenix/blob/173db6a2169c9fa6c4a6f49c7b46af7b47a92319/app/src/store/pointCloudStore.ts#L1159

Reply:
In my mind, loading prompts based on the value set by the prompt combobox _is_ a synchronization task.

Whenever the prompt id set on the instance changes we should make the instance template reflect the prompt.

Moving to something more imperative feels like it may lead to scenarios where changing the prompt does not update the instance. 

Let me think on it some more, I could be convinced away from this

Reply:
Removed in favor of fetchQuery callback